cumulusci/tasks/bulkdata/utils.py

```diff
@@ -13,15 +12,66 @@
 from cumulusci.utils import convert_to_snake_case
 
 
-def batch_iterator(iterator, n=10000):
-    while True:
-        batch = list(itertools.islice(iterator, n))
-        if not batch:
-            return
+def batch_iterator(iterator, fields, n=10000, char_limit=10000000):
+    """Given an iterator of records, return batches that (in order of precedence):
+    (1) Do not exceed the given character limit
+    (2) Contain no more than n records.
+    """
 
+    per_record_csv_chars = csv_const_chars_per_record(len(fields))
+    csv_field_chars = sum_chars(fields) + per_record_csv_chars
+
+    current_chars = 0
+    current_chars += csv_field_chars
+
+    batch = []
+    for record in iterator:
+        # Does the next record put us over the character limit?
+        if sum_chars(record) + per_record_csv_chars + current_chars > char_limit:
+            yield batch
+            batch = []
+            current_chars = csv_field_chars
+
+        batch.append(record)
+        current_chars += sum_chars(record) + per_record_csv_chars
+
+        # yield batch if we're at desired size
+        if len(batch) == n:
+            yield batch
+            batch = []
+            current_chars = csv_field_chars
+
+    # give back anything leftover
+    if batch:
         yield batch
 
 
+def csv_const_chars_per_record(num_fields):
+    """Given number of fields (or values) on a record, return the
+    number of **per record** constant characters that would be applied
+    in a .csv file format.
+
+    Given fields:
+        ['Id', 'FirstName', 'LastName']
+
+    The corresponding line in a csv file would be:
+        'Id','FirstName','LastName'\n
+
+    For each field there are:
+        quotations = num_fields * 2
+        commas = num_fields - 1
+        newlines = 1
+    Some algebra gives us: 3 * num_fields
+    #TODO: Do we include spaces in between values?
```

Comment:
Anyone know if we include spaces in the final csv data we send out?

Reply:
That behavior's owned by `csv`. It doesn't include spaces.

By default, a newline is `\r\n`, though, for 2 chars.

Reply:
Actually, there's something else nasty lurking here. `csv` will _quote_ specific field values, and will also escape embedded quotes in the field data. That will result in expansion of the field values we are sending. For some `csv.writer` `w`, 

    w.writerow(['test stuff', 'fo"o'])

yields

    "test\nstuff","fo""o"\r\n

I don't immediately see a way around this other than either walking across all the record values and applying the expected CSV transformations, which is a bit nasty, or serializing the entire CSV file to disk before uploading it, which is quite nasty and a major change to the architecture. Do you have any better thoughts?

Reply:
We use the default CSV dialect (https://docs.python.org/3/library/csv.html#csv-fmt-params).

Reply:
@davidmreed - Since the input/output of `batch_iterator()` has remained constant, this would be a bug that already exists in the code and isn't something being introduced correct? If you can confirm I can log a new bug record in GUS to address this.